### ISSUE TITLE
Substituter query parallelism is far too low

### ISSUE THREAD

**Issue by Kha** [NixOS/nix#5118]
TL;DR: a single-line change can make binary cache querying 50x as fast

Continuing from #5109, I decided to investigate *why* substitution (querying) is/seems to be fairly slow. Okay, maybe my 800 derivations are simply a bit much, I thought, but then I found @edolstra's [commit message](https://github.com/kha/nix/commit/90ad02bf626b885a5dd8967894e2eafc953bdf92) stating that HTTP/2 should handle this in under a second. So what's going on? The async API described in that commit never really got put to use as far as I can see, so it turns out that throughput is (*severely*) limited by the size of the [thread pool](https://github.com/kha/nix/blob/a6ba313a0aac3b6e2fef434cb42d190a0849238e/src/libstore/misc.cc#L103) used to fetch path infos in parallel via the synchronous API. The standard size of that pool is the number of hardware threads (8 on this machine), but we can improve throughput by >50x by massively scaling up that pool, at least in my use case!

#threads | "querying missing paths"
--- | ---
8 | 38s
32 | 10s
100 | 1.89s
300 | 0.89s
1000 | 0.73s

Now I can't talk about any substituters other than HTTP/2 binary caches and I don't know how exactly this should be fixed (should the pool size be a new option (with a high default or low default?), should the thread pool be replaced with true asynchrony, should doing the same for actual substitution be investigated as well, should substitution querying and actual substitution be merged after all, ...), but obviously I would be very happy about *any* solution removing this bottleneck and would be willing to implement it when there is consensus on the design.

**Comment by domenkozar** [comment]
Can you benchmark more entries between 8 and 100? That would yield a good default.

Without changing the design, I'd propose to chose a better default and make the pool size configurable.

From the Cachix side it would be best if Nix implemented batched lookups, that would probably require something like returning JSON with a list of narinfo entries. That would also require adding something like `DynamicQuerying: true` to `nix-cache-info`.

**Comment by michaelpj** [comment]
Possibly also relevant: https://github.com/NixOS/nix/issues/3019

**Comment by Kha** [comment]
> Can you benchmark more entries between 8 and 100? That would yield a good default.

I added another entry for 32 threads, though the benchmark should probably be redone to ensure consistent conditions. I'm a bit suspicious of the 100 threads entry now, its cumulative time (threads * total time) is much lower than the other entries. I've been recompiling Nix for each change in thread count so far :grimacing: .

> Without changing the design, I'd propose to chose a better default and make the pool size configurable.

That would definitely be the simplest solution, though the table shows that you are basically guaranteed to waste either time or resources. I would be very happy to see a thread-less solution eventually.

**Comment by Kha** [comment]
> Possibly also relevant: #3019

Yes, if your derivation graph is "too linear" (i.e. not enough derivations that are independent and thus can be queried in parallel) to saturate libcurl's queue, that could help as well.

**Comment by domenkozar** [comment]
@kha How about we bump the default to 32 and make it configurable?

**Comment by Kha** [comment]
Fine with me, I can open a PR

**Comment by domenkozar** [comment]
@Kha 2.4 is going out soon, maybe you can get this in time?

**Comment by domenkozar** [comment]
> Possibly also relevant: #3019

Also relevant: https://github.com/NixOS/nix/issues/3379

**Comment by DavHau** [comment]
There already is an option `http-connections`
https://github.com/NixOS/nix/blob/886ad0055f20ca70247b1290759903f244c27f88/src/libstore/filetransfer.hh#L20-L27

Why not just re-use that option and make it apply to substituter queries?

**Comment by Kha** [comment]
As demonstrated above, HTTP/2 querying can scale far beyond the number of TCP connections, which I did not touch. It's all about keeping all connections saturated for multiplexing I believe. I'm pretty sure we wouldn't want to open a thousand TCP connections, but a thousand threads is not really an issue (but still not the default in my PR).

**Comment by cab404** [comment]
I am wondering: are there any asynchronous IO options out there?

**Comment by nrdxp** [comment]
This may be a red-herring. I am working on improving the querying code and making it asychronous. I haven't opened a PR yet but I'm close. You can still see a version of the code I have here:
https://github.com/divnix/nix-uncached/blob/master/src/nix-uncached.cc#L109-L112

This alone is enough to drastically speed up the queries to the point that it takes more time to compute the closure that it does to query the cache, but there is a problem; I hit the connection limit of cloudflare several times with this implementation. This caused me to believe that connections are being closed too early, and indeed it appears a connection is opened and closed for each path.

This is essentially because `Store::queryValidPaths` calls `Store::queryPathInfo` per path, which in turn calls `Store::queryPathInfoUncached`, which openes a connection each time at the start of its work and presumably closed it when the connection goes out of scope.

We should optimize this code to keep the connection open for as long as possible. Perhaps we can simply add a `ConnectionHandle` on the `RemoteStore` class so that all these instance-method's can look for an open connection first, before starting a new one.

**Comment by Kha** [comment]
> [opens a connection](https://github.com/NixOS/nix/blob/189261c3449b602e979519844529d1adb2767f71/src/libstore/remote-store.cc#L443) each time

Look closer, the connection is taken from a pool. Also, this issue is mostly about `BinaryCacheStore`, not `RemoteStore`.

> We should still respect the connection limit

Does your code do that? I didn't see it immediately.

**Comment by nrdxp** [comment]
No, in-tree I'm using the asynchronous version of `Store::queryPathInfo` instead of using `std::async` and respecting the connection limit is one thing I need to add still. I guess I should post my code in a wip branch

**Comment by xokdvium** [comment]
Async path queries in https://github.com/NixOS/nix/pull/14786. No threads are really needed there, since most of the time we are just waiting on the narinfo transfer to complete.

**Comment by xokdvium** [comment]
Based on that async infra I've now also asyncified `queryMissing` that previously relied on threads for concurrency: https://github.com/NixOS/nix/pull/15474.
